### PR TITLE
Refix compact names w/o breaking local names

### LINKED ISSUE #1233: Zinc reports incorrect class file names
## Steps

```sh
git clone git@github.com:vasilmkd/zinc-wrong-class-names.git
cd zinc-wrong-class-names
sbt "clean; show compileOutputs"
```

**Warning:** The compilation can take 35+ seconds to complete. Check the note at the end of this post.

The output should look similar to the following, modulo local machine paths. Only the names of the `.class` files are important.

```
vasil@Vasil-JetBrains zinc-wrong-class-names % sbt "clean; show compileOutputs"
[info] welcome to sbt 1.9.3 (Eclipse Adoptium Java 17.0.8)
[info] loading global plugins from /Users/vasil/.sbt/1.0/plugins
[info] loading project definition from /Users/vasil/IdeaProjects/zinc-wrong-class-names/project
[info] loading settings for project zinc-wrong-class-names from build.sbt ...
[info] set current project to zinc-wrong-class-names (in build file:/Users/vasil/IdeaProjects/zinc-wrong-class-names/)
[success] Total time: 0 s, completed Aug 7, 2023, 5:47:03 PM
[info] compiling 1 Scala source to /Users/vasil/IdeaProjects/zinc-wrong-class-names/target/scala-2.13/classes ...
[info] * /Users/vasil/IdeaProjects/zinc-wrong-class-names/target/scala-2.13/classes/OuterLevelWithVeryVeryVeryLongClassName1$.class
[info] * /Users/vasil/IdeaProjects/zinc-wrong-class-names/target/scala-2.13/classes/OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVeryVeryVeryLongClassName2$.class
[info] * /Users/vasil/IdeaProjects/zinc-wrong-class-names/target/scala-2.13/classes/OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVeryVeryVeryLongClassName2$OuterLevelWithVeryVeryVeryLongClassName3$.class
[info] * /Users/vasil/IdeaProjects/zinc-wrong-class-names/target/scala-2.13/classes/OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVeryVeryVeryLongClassName2$OuterLevelWithVeryVeryVeryLongClassName3$OuterLevelWithVeryVeryVeryLongClassName4$.class
[info] * /Users/vasil/IdeaProjects/zinc-wrong-class-names/target/scala-2.13/classes/OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVeryVeryVeryLongClassName2$OuterLevelWithVeryVeryVeryLongClassName3$OuterLevelWithVeryVeryVeryLongClassName4$OuterLevelWithVeryVeryVeryLongClassName5$.class
[info] * /Users/vasil/IdeaProjects/zinc-wrong-class-names/target/scala-2.13/classes/OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVeryVeryVeryLongClassName2$OuterLevelWithVeryVeryVeryLongClassName3$OuterLevelWithVeryVeryVeryLongClassName4$OuterLevelWithVeryVeryVeryLongClassName5$OuterLevelWithVeryVeryVeryLongClassName6$.class
[info] * /Users/vasil/IdeaProjects/zinc-wrong-class-names/target/scala-2.13/classes/OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVeryVeryVeryLongClassName2$OuterLevelWithVeryVeryVeryLongClassName3$OuterLevelWithVeryVeryVeryLongClassName4$OuterLevelWithVeryVeryVeryLongClassName5$OuterLevelWithVeryVeryVeryLongClassName6$OuterLevelWithVeryVeryVeryLongClassName7$.class
[info] * /Users/vasil/IdeaProjects/zinc-wrong-class-names/target/scala-2.13/classes/OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVeryVeryVeryLongClassName2$OuterLevelWithVeryVeryVeryLongClassName3$OuterLevelWithVeryVeryVeryLongClassName4$OuterLevelWithVeryVeryVeryLongClassName5$OuterLevelWithVeryVeryVeryLongClassName6$OuterLevelWithVeryVeryVeryLongClassName7$OuterLevelWithVeryVeryVeryLongClassName8$.class
[info] * /Users/vasil/IdeaProjects/zinc-wrong-class-names/target/scala-2.13/classes/OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVeryVeryVeryLongClassName2$OuterLevelWithVeryVeryVeryLongClassName3$OuterLevelWithVeryVeryVeryLongClassName4$OuterLevelWithVeryVeryVeryLongClassName5$OuterLevelWithVeryVeryVeryLongClassName6$OuterLevelWithVeryVeryVeryLongClassName7$OuterLevelWithVeryVeryVeryLongClassName8$OuterLevelWithVeryVeryVeryLongClassName9$.class
[info] * /Users/vasil/IdeaProjects/zinc-wrong-class-names/target/scala-2.13/classes/OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVeryVeryVeryLongClassName2$OuterLevelWithVeryVeryVeryLongClassName3$OuterLevelWithVeryVeryVeryLongClassName4$OuterLevelWithVeryVeryVeryLongClassName5$OuterLevelWithVeryVeryVeryLongClassName6$OuterLevelWithVeryVeryVeryLongClassName7$OuterLevelWithVeryVeryVeryLongClassName8$OuterLevelWithVeryVeryVeryLongClassName9$OuterLevelWithVeryVeryVeryLongClassName10$.class
[info] * /Users/vasil/IdeaProjects/zinc-wrong-class-names/target/scala-2.13/classes/OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVeryVeryVeryLongClassName2$OuterLevelWithVeryVeryVeryLongClassName3$OuterLevelWithVeryVeryVeryLongClassName4$OuterLevelWithVeryVeryVeryLongClassName5$OuterLevelWithVeryVeryVeryLongClassName6$OuterLevelWithVeryVeryVeryLongClassName7$OuterLevelWithVeryVeryVeryLongClassName8$OuterLevelWithVeryVeryVeryLongClassName9$OuterLevelWithVeryVeryVeryLongClassName10$OuterLevelWithVeryVeryVeryLongClassName11$.class
[info] * /Users/vasil/IdeaProjects/zinc-wrong-class-names/target/scala-2.13/classes/OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVeryVeryVeryLongClassName2$OuterLevelWithVeryVeryVeryLongClassName3$OuterLevelWithVeryVeryVeryLongClassName4$OuterLevelWithVeryVeryVeryLongClassName5$OuterLevelWithVeryVeryVeryLongClassName6$OuterLevelWithVeryVeryVeryLongClassName7$OuterLevelWithVeryVeryVeryLongClassName8$OuterLevelWithVeryVeryVeryLongClassName9$OuterLevelWithVeryVeryVeryLongClassName10$OuterLevelWithVeryVeryVeryLongClassName11$OuterLevelWithVeryVeryVeryLongClassName12$.class
[info] * /Users/vasil/IdeaProjects/zinc-wrong-class-names/target/scala-2.13/classes/OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVeryVeryVeryLongClassName2$OuterLevelWithVeryVeryVeryLongClassName3$OuterLevelWithVeryVeryVeryLongClassName4$OuterLevelWithVeryVeryVeryLongClassName5$OuterLevelWithVeryVeryVeryLongClassName6$OuterLevelWithVeryVeryVeryLongClassName7$OuterLevelWithVeryVeryVeryLongClassName8$OuterLevelWithVeryVeryVeryLongClassName9$OuterLevelWithVeryVeryVeryLongClassName10$OuterLevelWithVeryVeryVeryLongClassName11$OuterLevelWithVeryVeryVeryLongClassName12$OuterLevelWithVeryVeryVeryLongClassName13$.class
[info] * /Users/vasil/IdeaProjects/zinc-wrong-class-names/target/scala-2.13/classes/OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVeryVeryVeryLongClassName2$OuterLevelWithVeryVeryVeryLongClassName3$OuterLevelWithVeryVeryVeryLongClassName4$OuterLevelWithVeryVeryVeryLongClassName5$OuterLevelWithVeryVeryVeryLongClassName6$OuterLevelWithVeryVeryVeryLongClassName7$OuterLevelWithVeryVeryVeryLongClassName8$OuterLevelWithVeryVeryVeryLongClassName9$OuterLevelWithVeryVeryVeryLongClassName10$OuterLevelWithVeryVeryVeryLongClassName11$OuterLevelWithVeryVeryVeryLongClassName12$OuterLevelWithVeryVeryVeryLongClassName13$OuterLevelWithVeryVeryVeryLongClassName14$.class
[info] * /Users/vasil/IdeaProjects/zinc-wrong-class-names/target/scala-2.13/classes/OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVeryVeryVeryLongClassName2$OuterLevelWithVeryVeryVeryLongClassName3$OuterLevelWithVeryVeryVeryLongClassName4$OuterLevelWithVeryVeryVeryLongClassName5$OuterLevelWithVeryVeryVeryLongClassName6$OuterLevelWithVeryVeryVeryLongClassName7$OuterLevelWithVeryVeryVeryLongClassName8$OuterLevelWithVeryVeryVeryLongClassName9$OuterLevelWithVeryVeryVeryLongClassName10$OuterLevelWithVeryVeryVeryLongClassName11$OuterLevelWithVeryVeryVeryLongClassName12$OuterLevelWithVeryVeryVeryLongClassName13$OuterLevelWithVeryVeryVeryLongClassName14$OuterLevelWithVeryVeryVeryLongClassName15$.class
[info] * /Users/vasil/IdeaProjects/zinc-wrong-class-names/target/scala-2.13/classes/OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVeryVeryVeryLongClassName2$OuterLevelWithVeryVeryVeryLongClassName3$OuterLevelWithVeryVeryVeryLongClassName4$OuterLevelWithVeryVeryVeryLongClassName5$OuterLevelWithVeryVeryVeryLongClassName6$OuterLevelWithVeryVeryVeryLongClassName7$OuterLevelWithVeryVeryVeryLongClassName8$OuterLevelWithVeryVeryVeryLongClassName9$OuterLevelWithVeryVeryVeryLongClassName10$OuterLevelWithVeryVeryVeryLongClassName11$OuterLevelWithVeryVeryVeryLongClassName12$OuterLevelWithVeryVeryVeryLongClassName13$OuterLevelWithVeryVeryVeryLongClassName14$OuterLevelWithVeryVeryVeryLongClassName15$OuterLevelWithVeryVeryVeryLongClassName16$.class
[info] * /Users/vasil/IdeaProjects/zinc-wrong-class-names/target/scala-2.13/classes/OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVeryVeryVeryLongClassName2$OuterLevelWithVeryVeryVeryLongClassName3$OuterLevelWithVeryVeryVeryLongClassName4$OuterLevelWithVeryVeryVeryLongClassName5$OuterLevelWithVeryVeryVeryLongClassName6$OuterLevelWithVeryVeryVeryLongClassName7$OuterLevelWithVeryVeryVeryLongClassName8$OuterLevelWithVeryVeryVeryLongClassName9$OuterLevelWithVeryVeryVeryLongClassName10$OuterLevelWithVeryVeryVeryLongClassName11$OuterLevelWithVeryVeryVeryLongClassName12$OuterLevelWithVeryVeryVeryLongClassName13$OuterLevelWithVeryVeryVeryLongClassName14$OuterLevelWithVeryVeryVeryLongClassName15$OuterLevelWithVeryVeryVeryLongClassName16$OuterLevelWithVeryVeryVeryLongClassName17$.class
[info] * /Users/vasil/IdeaProjects/zinc-wrong-class-names/target/scala-2.13/classes/OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVeryVeryVeryLongClassName2$OuterLevelWithVeryVeryVeryLongClassName3$OuterLevelWithVeryVeryVeryLongClassName4$OuterLevelWithVeryVeryVeryLongClassName5$OuterLevelWithVeryVeryVeryLongClassName6$OuterLevelWithVeryVeryVeryLongClassName7$OuterLevelWithVeryVeryVeryLongClassName8$OuterLevelWithVeryVeryVeryLongClassName9$OuterLevelWithVeryVeryVeryLongClassName10$OuterLevelWithVeryVeryVeryLongClassName11$OuterLevelWithVeryVeryVeryLongClassName12$OuterLevelWithVeryVeryVeryLongClassName13$OuterLevelWithVeryVeryVeryLongClassName14$OuterLevelWithVeryVeryVeryLongClassName15$OuterLevelWithVeryVeryVeryLongClassName16$OuterLevelWithVeryVeryVeryLongClassName17$OuterLevelWithVeryVeryVeryLongClassName18$.class
[info] * /Users/vasil/IdeaProjects/zinc-wrong-class-names/target/scala-2.13/classes/OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVeryVeryVeryLongClassName2$OuterLevelWithVeryVeryVeryLongClassName3$OuterLevelWithVeryVeryVeryLongClassName4$OuterLevelWithVeryVeryVeryLongClassName5$OuterLevelWithVeryVeryVeryLongClassName6$OuterLevelWithVeryVeryVeryLongClassName7$OuterLevelWithVeryVeryVeryLongClassName8$OuterLevelWithVeryVeryVeryLongClassName9$OuterLevelWithVeryVeryVeryLongClassName10$OuterLevelWithVeryVeryVeryLongClassName11$OuterLevelWithVeryVeryVeryLongClassName12$OuterLevelWithVeryVeryVeryLongClassName13$OuterLevelWithVeryVeryVeryLongClassName14$OuterLevelWithVeryVeryVeryLongClassName15$OuterLevelWithVeryVeryVeryLongClassName16$OuterLevelWithVeryVeryVeryLongClassName17$OuterLevelWithVeryVeryVeryLongClassName18$OuterLevelWithVeryVeryVeryLongClassName19$.class
[info] * /Users/vasil/IdeaProjects/zinc-wrong-class-names/target/scala-2.13/classes/OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVeryVeryVeryLongClassName2$OuterLevelWithVeryVeryVeryLongClassName3$OuterLevelWithVeryVeryVeryLongClassName4$OuterLevelWithVeryVeryVeryLongClassName5$OuterLevelWithVeryVeryVeryLongClassName6$OuterLevelWithVeryVeryVeryLongClassName7$OuterLevelWithVeryVeryVeryLongClassName8$OuterLevelWithVeryVeryVeryLongClassName9$OuterLevelWithVeryVeryVeryLongClassName10$OuterLevelWithVeryVeryVeryLongClassName11$OuterLevelWithVeryVeryVeryLongClassName12$OuterLevelWithVeryVeryVeryLongClassName13$OuterLevelWithVeryVeryVeryLongClassName14$OuterLevelWithVeryVeryVeryLongClassName15$OuterLevelWithVeryVeryVeryLongClassName16$OuterLevelWithVeryVeryVeryLongClassName17$OuterLevelWithVeryVeryVeryLongClassName18$OuterLevelWithVeryVeryVeryLongClassName19$OuterLevelWithVeryVeryVeryLongClassName20$.class
[info] * /Users/vasil/IdeaProjects/zinc-wrong-class-names/target/scala-2.13/classes/OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVeryVeryVeryLongClassName2$OuterLevelWithVeryVeryVeryLongClassName3$OuterLevelWithVeryVeryVeryLongClassName4$OuterLevelWithVeryVeryVeryLongClassName5$OuterLevelWithVeryVeryVeryLongClassName6$OuterLevelWithVeryVeryVeryLongClassName7$OuterLevelWithVeryVeryVeryLongClassName8$OuterLevelWithVeryVeryVeryLongClassName9$OuterLevelWithVeryVeryVeryLongClassName10$OuterLevelWithVeryVeryVeryLongClassName11$OuterLevelWithVeryVeryVeryLongClassName12$OuterLevelWithVeryVeryVeryLongClassName13$OuterLevelWithVeryVeryVeryLongClassName14$OuterLevelWithVeryVeryVeryLongClassName15$OuterLevelWithVeryVeryVeryLongClassName16$OuterLevelWithVeryVeryVeryLongClassName17$OuterLevelWithVeryVeryVeryLongClassName18$OuterLevelWithVeryVeryVeryLongClassName19$OuterLevelWithVeryVeryVeryLongClassName20$MalformedNameExample$.class
[info] * /Users/vasil/IdeaProjects/zinc-wrong-class-names/target/scala-2.13/classes/OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVeryVeryVeryLongClassName2$OuterLevelWithVeryVeryVeryLongClassName3$OuterLevelWithVeryVeryVeryLongClassName4$OuterLevelWithVeryVeryVeryLongClassName5$OuterLevelWithVeryVeryVeryLongClassName6$OuterLevelWithVeryVeryVeryLongClassName7$OuterLevelWithVeryVeryVeryLongClassName8$OuterLevelWithVeryVeryVeryLongClassName9$OuterLevelWithVeryVeryVeryLongClassName10$OuterLevelWithVeryVeryVeryLongClassName11$OuterLevelWithVeryVeryVeryLongClassName12$OuterLevelWithVeryVeryVeryLongClassName13$OuterLevelWithVeryVeryVeryLongClassName14$OuterLevelWithVeryVeryVeryLongClassName15$OuterLevelWithVeryVeryVeryLongClassName16$OuterLevelWithVeryVeryVeryLongClassName17$OuterLevelWithVeryVeryVeryLongClassName18$OuterLevelWithVeryVeryVeryLongClassName19$OuterLevelWithVeryVeryVeryLongClassName20$MalformedNameExample.class
[info] * /Users/vasil/IdeaProjects/zinc-wrong-class-names/target/scala-2.13/classes/OuterLevelWithVeryVeryVeryLongClassName1.class
[info] * /Users/vasil/IdeaProjects/zinc-wrong-class-names/target/scala-2.13/zinc/inc_compile_2.13.zip
[success] Total time: 35 s, completed Aug 7, 2023, 5:47:38 PM
```

## Problem

These names would never be generated by the `scalac` compiler, because they are too long for the local file systems and can cause exceptions when opened for reading in a JVM process. They exceed 255 characters in length.

```sh
cd target/scala-2.13/classes
ls
```

These are the actual `.class` files generated by `scalac`.

```
OuterLevelWithVeryVeryVeryLongClassName1$.class
OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVer$$$$21cd1be25088dedb16ca1b3c346466fb$$$$eryLongClassName7$OuterLevelWithVeryVeryVeryLongClassName8$.class
OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVer$$$$21cd1be25088dedb16ca1b3c346466fb$$$$eryLongClassName7$OuterLevelWithVeryVeryVeryLongClassName8$OuterLevelWithVeryVeryVeryLongClassName9$.class
OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVer$$$$361c0db7030df22c21772c193d391$$$$yLongClassName19$OuterLevelWithVeryVeryVeryLongClassName20$.class
OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVer$$$$361c0db7030df22c21772c193d391$$$$yLongClassName19$OuterLevelWithVeryVeryVeryLongClassName20$MalformedNameExample$.class
OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVer$$$$361c0db7030df22c21772c193d391$$$$yLongClassName19$OuterLevelWithVeryVeryVeryLongClassName20$MalformedNameExample.class
OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVer$$$$7e70169f8f612a14697f8fd52119750$$$$yLongClassName15$OuterLevelWithVeryVeryVeryLongClassName16$.class
OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVer$$$$7e70169f8f612a14697f8fd52119750$$$$yLongClassName15$OuterLevelWithVeryVeryVeryLongClassName16$OuterLevelWithVeryVeryVeryLongClassName17$.class
OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVer$$$$973a10bc1cd9792c988bf49bee47b56$$$$ryLongClassName9$OuterLevelWithVeryVeryVeryLongClassName10$.class
OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVer$$$$973a10bc1cd9792c988bf49bee47b56$$$$ryLongClassName9$OuterLevelWithVeryVeryVeryLongClassName10$OuterLevelWithVeryVeryVeryLongClassName11$.class
OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVer$$$$9a7e92c93110f2442b3ebd9c453788$$$$yLongClassName11$OuterLevelWithVeryVeryVeryLongClassName12$.class
OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVer$$$$9a7e92c93110f2442b3ebd9c453788$$$$yLongClassName11$OuterLevelWithVeryVeryVeryLongClassName12$OuterLevelWithVeryVeryVeryLongClassName13$.class
OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVer$$$$9f428b41847a5866a8323f5944a36d$$$$yLongClassName13$OuterLevelWithVeryVeryVeryLongClassName14$.class
OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVer$$$$9f428b41847a5866a8323f5944a36d$$$$yLongClassName13$OuterLevelWithVeryVeryVeryLongClassName14$OuterLevelWithVeryVeryVeryLongClassName15$.class
OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVer$$$$d292b257125352e68f2f62fd843b554$$$$yLongClassName17$OuterLevelWithVeryVeryVeryLongClassName18$.class
OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVer$$$$d292b257125352e68f2f62fd843b554$$$$yLongClassName17$OuterLevelWithVeryVeryVeryLongClassName18$OuterLevelWithVeryVeryVeryLongClassName19$.class
OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVer$$$$f1136b7659b86ea0804013acf91cbef$$$$eryLongClassName5$OuterLevelWithVeryVeryVeryLongClassName6$.class
OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVer$$$$f1136b7659b86ea0804013acf91cbef$$$$eryLongClassName5$OuterLevelWithVeryVeryVeryLongClassName6$OuterLevelWithVeryVeryVeryLongClassName7$.class
OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVeryVeryVeryLongClassName2$.class
OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVeryVeryVeryLongClassName2$OuterLevelWithVeryVeryVeryLongClassName3$.class
OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVeryVeryVeryLongClassName2$OuterLevelWithVeryVeryVeryLongClassName3$OuterLevelWithVeryVeryVeryLongClassName4$.class
OuterLevelWithVeryVeryVeryLongClassName1$OuterLevelWithVeryVeryVeryLongClassName2$OuterLevelWithVeryVeryVeryLongClassName3$OuterLevelWithVeryVeryVeryLongClassName4$OuterLevelWithVeryVeryVeryLongClassName5$.class
OuterLevelWithVeryVeryVeryLongClassName1.class
```

These names come from https://github.com/scala/scala/blob/98a42345385cc8adccde36c21f3192d4b4862087/src/reflect/scala/reflect/internal/StdNames.scala#L49-L86.

## Expectation

The `compileOutputs` reported by Zinc should match the actual class files generated by `scalac`.

In IntelliJ IDEA, this problem manifests as part of our compilation pipeline which uses Zinc.
https://youtrack.jetbrains.com/issue/SCL-21477/File-name-too-long-error-when-building-projects-like-Apache-Spark

This is a pretty serious issue. I would appreciate any help in diagnosing it.

## Notes
On a side note, the compilation of the single Scala source file `main.scala` takes 35+ seconds and cannot be canceled with `ctrl+C`. This seems excessive, since most of the definitions are empty object definitions.

### PR DESCRIPTION
**Pull request by dwijnand**
The previous fix phase-traveled to after flatten, which is the threshold
for Symbol#name to start returning flattened names.  Unfortunately there
is a bug (one of a few in nsc) in that the owner of the local class
symbols isn't changed as a part of a lambda lift info transform, it's
changed as a side-effect of doing lambda lift tree transform.
Additionally the flattened name is cached on first use, which in this
case means using the wrong owner - that's why the compiler failed in
pekko, because there's a check in the backend that the same classfile
(by name) isn't emitted twice.

Turns out, we don't need the full names of method local classes, so we
can push that condition a bit higher in registerGeneratedClasses,
avoiding useless work, useless phase travel, and erroneous caching.
We also just straight up avoid phase traveling from xsbt-api (which runs
after typer) all the way to flatten (which is almost at the end) for the
flatten, compactified name, by reimplementing `flattenedName`, with the
correct, recursive, flattened owner and name.

By the by, the test case is pretty convoluted, despite trying to make it
simpler - how many times have you defined and instantiated a
method-local anonymous class, defining one of its methods using a case
object?

### CODE REVIEW

File: internal/zinc-scripted/src/test/scala/sbt/internal/inc/IncHandler.scala [hunk 1]
```diff
@@ -829,6 +836,14 @@ case class ProjectStructure(
       }
       ()
     }
+
+  def assertShort(assertion: Boolean, message: => Any) = {
```
Comment (eed3si9n) [inline internal/zinc-scripted/src/test/scala/sbt/internal/inc/IncHandler.scala:840]: nit: is this in use?

Comment (dwijnand) [reply internal/zinc-scripted/src/test/scala/sbt/internal/inc/IncHandler.scala:840]: Not any more.  I was leaving it for the next time, if that's ok - or I can remove it.

File: internal/compiler-bridge/src/main/scala/scala/reflect/ReflectAccess.scala [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+  def flattenedName(g: Global)(sym: g.Symbol): g.Name = {
+    val owner = sym.owner
+
+    val flat =
+      if (
+        owner.hasPackageFlag ||
+        owner.isRoot ||
+        owner == g.NoSymbol ||
+        owner.hasPackageFlag
```
Comment (lrytz) [inline internal/compiler-bridge/src/main/scala/scala/reflect/ReflectAccess.scala]: same as line 17 :)

File: internal/compiler-bridge/src/main/scala/xsbt/CallbackGlobal.scala [hunk 3]
```diff
@@ -225,19 +226,21 @@ sealed class ZincCompiler(settings: Settings, dreporter: DelegatingReporter, out
...
       val nSize = encodedName.length - (if (symName.endsWith(nme.LOCAL_SUFFIX_STRING)) 1 else 0)
-      if (sym.isRoot || sym.isRootPackage || sym == NoSymbol || sym.owner.isEffectiveRoot) {
+      val owner = flattenedOwner(this)(sym)
+      if (sym.isRoot || sym.isRootPackage || sym == NoSymbol || owner.isEffectiveRoot) {
         val capacity = size + nSize
         b = new java.lang.StringBuffer(capacity)
         b.append(chrs, encodedName.start, nSize)
       } else {
-        val next = if (sym.owner.isPackageObjectClass) sym.owner else sym.effectiveOwner.enclClass
+        val next = if (owner.isPackageObjectClass) owner else owner.skipPackageObject.enclClass
```
Comment (lrytz) [inline internal/compiler-bridge/src/main/scala/xsbt/CallbackGlobal.scala]: I guess `enclClass` is redundant now, as `flattenedOwner` walks to the package (or package object) from where all owners will be packages where `enclClass` is identity (PackageClassSymbol)?

File: internal/compiler-bridge/src/main/scala/xsbt/CallbackGlobal.scala [hunk 4]
```diff
@@ -225,19 +226,21 @@ sealed class ZincCompiler(settings: Settings, dreporter: DelegatingReporter, out
...
         loop(size + nSize + 1, next)
         // Addition to normal `fullName` to produce correct names for nested non-local classes
-        if (sym.isNestedClass) b.append(nme.MODULE_SUFFIX_STRING) else b.append(separator)
+        val isNestedClass = sym.isClass && !owner.isPackageClass
```
Comment (lrytz) [inline internal/compiler-bridge/src/main/scala/xsbt/CallbackGlobal.scala]: The only remaining case is now if `owner` is a package object, right? Should we change it to `owner.isPackageObjectClass`?

**Review by lrytz — APPROVED** [review]
LGTM now!

**Review by eed3si9n — APPROVED** [review]
ok to land, but I'll probably ship this as 1.10.x.

**Pull request merged**